src/webgl/p5.Shader.js

```diff
@@ -0,0 +1,330 @@
+/**
+ * This module defines the p5.Shader class
+ * @module Lights, Camera
+ * @submodule Shaders
+ * @for p5
+ * @requires core
+ */
+
+var p5 = require('../core/core');
+
+/**
+ * Shader class for WEBGL Mode
+ * @class p5.Shader
+ * @constructor
+ * @param {renderer} [p5.Renderer.GL] an instance of p5.RendererGL that
+ * will provide the GL context for this new p5.Shader
+ * @param {vertSrc} [String] source code for the vertex shader (as a string)
+ * @param {fragSrc} [String] source code for the fragment shader (as a string)
+ */
+p5.Shader = function(renderer, vertSrc, fragSrc) {
+  // TODO: adapt this to not take ids, but rather,
+  // to take the source for a vertex and fragment shader
+  // to enable custom shaders at some later date
+  this._renderer = renderer;
+  this._vertSrc = vertSrc;
+  this._fragSrc = fragSrc;
+  this._vertShader = -1;
+  this._fragShader = -1;
+  this.glProgram = 0;
+  this.loadedAttributes = false;
+  this.attributes = {};
+  this.loadedUniforms = false;
+  this.uniforms = {};
+  this.bound = false;
```

Comment:
Yes. My thinking is that some of these are in-between: that they are something more akin to what Java would call protected (where the renderer, and components within the p5 GL pipeline, have access to them and are expected to use them but other modules within the project would not). I don't know if there are JS conventions for this...